core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,259 @@
+     */
+    private static List<Subchannel> filterFailingSubchannels(Collection<Subchannel> subchannels) {
+      List<Subchannel> nonFailingSubchannels = Lists.newArrayList();
+      for (Subchannel subchannel : subchannels) {
+        if (getSubchannelState(subchannel) != TRANSIENT_FAILURE) {
+          nonFailingSubchannels.add(subchannel);
+        }
+      }
+      return nonFailingSubchannels;
+    }
+
+    /**
+     * Converts list of {@link ResolvedServerInfoGroup} to {@link EquivalentAddressGroup} set.
+     */
+    private static Set<EquivalentAddressGroup> resolvedServerInfoGroupToEquivalentAddressGroup(
+        List<ResolvedServerInfoGroup> groupList) {
+      Set<EquivalentAddressGroup> addrs = new HashSet<EquivalentAddressGroup>(groupList.size());
+      for (ResolvedServerInfoGroup group : groupList) {
+        for (ResolvedServerInfo server : group.getResolvedServerInfoList()) {
+          addrs.add(server.toEquivalentAddressGroup());
+        }
+      }
+      return addrs;
+    }
+
+    /**
+     * If all subchannels are TRANSIENT_FAILURE, return the Status associated with an arbitrary
+     * subchannel otherwise, return null.
+     */
+    @Nullable
+    private Status getAggregatedError() {
+      for (Subchannel subchannel : subchannels.values()) {
+        if (subchannel.getAttributes().get(STATE_INFO).get().getState() != TRANSIENT_FAILURE) {
+          return null;
+        }
+      }
+      return getSubchannelStatus(subchannels.values().iterator().next());
+    }
+
+    @VisibleForTesting
+    Collection<Subchannel> getSubchannels() {
+      return subchannels.values();
+    }
+
+    private static ConnectivityState getSubchannelState(Subchannel subchannel) {
+      return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO").get()
+          .getState();
+    }
+
+    private static void setSubchannelStateInfo(Subchannel subchannel,
+        ConnectivityStateInfo newStateInfo) {
+      checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO").set(newStateInfo);
+    }
+
+    private static Status getSubchannelStatus(Subchannel subchannel) {
+      return checkNotNull(subchannel.getAttributes().get(STATE_INFO), "STATE_INFO").get()
+          .getStatus();
+    }
+  }
+
+  @VisibleForTesting
+  static class Picker extends SubchannelPicker {
+    @GuardedBy("this")
+    final Iterator<Subchannel> subchannelIterator;
+    @Nullable
+    final Status status;
+
+    Picker(Iterable<Subchannel> list, @Nullable Status status) {
+      this.subchannelIterator = Iterables.cycle(list).iterator();
+      this.status = status;
+    }
+
+    @Override
+    public PickResult pickSubchannel(Attributes affinity, Metadata headers) {
+      // This will be true only if it's empty
+      if (subchannelIterator.hasNext()) {
+        return PickResult.withSubchannel(subchannelIterator.next());
+      } else {
+        return PickResult.withError(checkNotNull(status, "status"));
```

Comment:
i'm not sure, but can't we check that `status != null` in the constructor if `subchannelIterator` is empty?